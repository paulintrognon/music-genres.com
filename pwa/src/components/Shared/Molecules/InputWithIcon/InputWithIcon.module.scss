.container {
  position: relative;
  display: flex;
  align-items: center;
}

.icon {
  position: absolute;
  width: auto;
  right: 1rem;
}
