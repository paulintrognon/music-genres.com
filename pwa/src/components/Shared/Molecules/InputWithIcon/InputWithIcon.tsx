import TextInput, { Props as TextInputProps } from '../../Atoms/TextInput/TextInput'
import classnames from 'classnames'
import classes from './InputWithIcon.module.scss'

interface Props extends TextInputProps {
  icon: React.ReactNode
  onSubmit?: () => void
}
const InputWithIcon: React.FC<Props> = ({ icon, className, onSubmit, ...props }) => {
  const handleOnSubmit = (event: React.FormEvent): void => {
    event.preventDefault()
    onSubmit && onSubmit()
  }

  return (
    <form className={classnames(classes.container, className)} onSubmit={handleOnSubmit}>
      <TextInput {...props} />
      <button type="submit" className={classes.icon}>
        {icon}
      </button>
    </form>
  )
}
export default InputWithIcon
