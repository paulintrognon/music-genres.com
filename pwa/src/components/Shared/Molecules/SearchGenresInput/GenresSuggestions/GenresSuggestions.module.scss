.suggestion {
  display: block;
  width: 100%;
  padding: 0.75rem 1rem;
  background-color: var(--color-secondary);
  
  a {
    text-decoration: underline;
  }
}

.hover:hover {
  background-color: var(--color-secondary-light);
  box-shadow: 0 1px 10px 0 rgba(0,0,0,.19);
  position: relative;
}