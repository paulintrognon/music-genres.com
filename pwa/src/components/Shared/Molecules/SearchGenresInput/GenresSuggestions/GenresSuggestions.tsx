import Link from 'next/link'
import { MusicGenreBaseType } from '../../../../../types/MusicGenre/MusicGenreBaseType'
import classnames from 'classnames'
import classes from './GenresSuggestions.module.scss'
import Loader from '../../../Atoms/Loader/Loader'

type Props = {
  isLoading: boolean
  genreName: string
  suggestions: MusicGenreBaseType[]
  slugToHref: (slug: string) => string
}
const GenresSuggestions: React.FC<Props> = ({ isLoading, genreName, suggestions, slugToHref }) => {
  if (isLoading) {
    return (
      <div className={classes.suggestion} style={{ textAlign: 'center' }}>
        <Loader />
      </div>
    )
  }
  if (suggestions.length === 0) {
    return (
      <div className={classes.suggestion}>
        &quot; <i>{genreName}</i> &quot; does not exist yet.{' '}
        <Link href={`/add/video/${genreName}`}>
          <a>Create it!</a>
        </Link>
      </div>
    )
  }
  return (
    <div>
      {suggestions.map((suggestion) => (
        <Link key={suggestion.slug} href={slugToHref(suggestion.slug)}>
          <a className={classnames(classes.suggestion, classes.hover)}>{suggestion.name}</a>
        </Link>
      ))}
    </div>
  )
}
export default GenresSuggestions
