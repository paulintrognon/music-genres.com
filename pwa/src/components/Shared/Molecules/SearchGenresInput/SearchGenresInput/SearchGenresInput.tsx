import { ChangeEvent, useState } from 'react'
import useSWR from 'swr'
import { GenreSuggestionsApiResultType } from '../../../../../types/Api/GenreSuggestionsApiResultType'
import MagnifyingGlassIcon from '../../../Atoms/Icons/MagnifyingGlass/MagnifyingGlass'
import InputWithIcon from '../../InputWithIcon/InputWithIcon'
import GenresSuggestions from '../GenresSuggestions/GenresSuggestions'
import classes from './SearchGenresInput.module.scss'

interface Props {
  slugToHref: (slug: string) => string
  onChange?: (entry: string) => void
  onSubmit?: (entry: string) => void
  onDirty?: () => void
  onClean?: () => void
}
const SearchGenresInput: React.FC<Props> = ({
  slugToHref,
  onChange,
  onSubmit,
  onDirty,
  onClean,
}) => {
  /**
   * On string query changes, we update query state
   */
  const [typedText, setTypedText] = useState('')
  const handleOnChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setTypedText(event.target.value)
    onChange && onChange(event.target.value)
  }

  /**
   * On query state changes, we fetch up to 5 genres suggestions
   */
  const { data, isValidating } = useSWR<GenreSuggestionsApiResultType>(
    typedText && `/api/music-genres/search?query=${typedText}&limit=5`
  )

  /**
   * On submit, we pass the typed text to onSubmit handler (if exists)
   */
  const handleOnSubmit = (): void => {
    onSubmit && onSubmit(typedText)
  }

  return (
    <div className={classes.container}>
      <InputWithIcon
        className={classes.searchInput}
        placeholder="Start looking for a genre"
        icon={<MagnifyingGlassIcon />}
        onDirty={onDirty}
        onClean={onClean}
        onChange={handleOnChange}
        onSubmit={handleOnSubmit}
      />
      {typedText ? (
        <GenresSuggestions
          isLoading={isValidating}
          genreName={typedText}
          suggestions={data?.result || []}
          slugToHref={slugToHref}
        />
      ) : null}
    </div>
  )
}
export default SearchGenresInput
