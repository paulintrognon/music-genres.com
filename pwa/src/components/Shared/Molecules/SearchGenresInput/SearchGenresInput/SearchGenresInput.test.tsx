import { shallow } from 'enzyme'
import SearchGenresInput from './SearchGenresInput'
import InputWithIcon from '../../InputWithIcon/InputWithIcon'
import GenresSuggestions from '../GenresSuggestions/GenresSuggestions'

// Mocking useSWR
import useSWR from 'swr'
jest.mock('swr')
const useSWRMock = useSWR as jest.Mock

const slugToHref = (slug: string): string => `/path/to/${slug}`

beforeEach(() => {
  jest.resetAllMocks()
  useSWRMock.mockReturnValue({ data: { result: [] } })
})

it('should render SearchGenres', () => {
  const wrapper = shallow(<SearchGenresInput slugToHref={slugToHref} />)
  expect(wrapper).toMatchSnapshot()
})

it('should search genres on input', async () => {
  const onDirty = jest.fn()
  const onClean = jest.fn()

  const wrapper = shallow(
    <SearchGenresInput onDirty={onDirty} onClean={onClean} slugToHref={slugToHref} />
  )

  // On component mount, useSWR should be called with an empty string
  expect(useSWRMock).toHaveBeenCalledTimes(1)
  expect(useSWRMock.mock.calls[0]).toEqual([''])
  expect(wrapper.find(GenresSuggestions)).toHaveLength(0)

  // On query "test" typed ...
  const suggestionsResult = [{ name: 'Genre', slug: 'genre' }]
  useSWRMock.mockReturnValueOnce({ data: { result: suggestionsResult } })
  wrapper.find(InputWithIcon).simulate('change', {
    target: { value: 'test' },
  })
  // ...useSWR should be called with corresponding API url
  expect(useSWRMock).toHaveBeenCalledTimes(2)
  expect(useSWRMock.mock.calls[1]).toEqual(['/api/music-genres/search?query=test&limit=5'])
  // ...and suggestions should be shown
  expect(wrapper).toMatchSnapshot()
})

it('should call onSubmit with typed text on submit', async () => {
  const onSubmit = jest.fn()

  const wrapper = shallow(<SearchGenresInput onSubmit={onSubmit} slugToHref={slugToHref} />)

  // On query "BardCore" typed ...
  wrapper.find(InputWithIcon).simulate('change', {
    target: { value: 'BardCore' },
  })

  // And on submit...
  wrapper.find(InputWithIcon).simulate('submit')

  // onSubmit should be called with "BardCode"
  expect(onSubmit).toHaveBeenCalledTimes(1)
  expect(onSubmit.mock.calls[0]).toEqual(['BardCore'])
})
