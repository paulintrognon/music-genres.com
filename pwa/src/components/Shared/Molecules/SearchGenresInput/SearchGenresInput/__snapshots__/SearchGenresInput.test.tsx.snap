// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchGenres 1`] = `
<div
  className="container"
>
  <InputWithIcon
    className="searchInput"
    icon={<MagnifyingGlassIcon />}
    onChange={[Function]}
    onSubmit={[Function]}
    placeholder="Start looking for a genre"
  />
</div>
`;

exports[`should search genres on input 1`] = `
<div
  className="container"
>
  <InputWithIcon
    className="searchInput"
    icon={<MagnifyingGlassIcon />}
    onChange={[Function]}
    onClean={[MockFunction]}
    onDirty={[MockFunction]}
    onSubmit={[Function]}
    placeholder="Start looking for a genre"
  />
  <GenresSuggestions
    genreName="test"
    slugToHref={[Function]}
    suggestions={
      Array [
        Object {
          "name": "Genre",
          "slug": "genre",
        },
      ]
    }
  />
</div>
`;
